Added test framework

mbivolan opened the pull request:
Changes:
- Added test framework
- Changed current implementation for the LISAv2 powershell tests to support the framework
- Changed 2 python test scripts to support the same parameter interface as bash test scripts
- Imported first sprint of Core tests from LISA

juhlee-microsoft commented on Testscripts/Windows/BVT-VERIFY-PRIVATEKEY-AUTHENTICATION.ps1:
```diff
@@ -1,85 +1,64 @@
-$resultArr = @()
-$isDeployed = DeployVMS -setupType $currentTestData.setupType -Distro $Distro -xmlConfig $xmlConfig
-#$isDeployed = "ICA-BVTDeployment-UbuntuCAPT-6-25-5-8-21"
-if ($isDeployed)
```
$isDeployed removed. Again I recall it was used in Azure before.

adriansuhov replied:
It is still used. But all "Deploy" code was moved to Libraries/RunTest.psm1. Also all checks are done there. It is used for both Azure and Hyper-V deployments. The difference is that on Azure we have one deployment/test (functionality was not changed) and on Hyper-V we have one deployment for all tests. Since the deployments are done in RunTest.psm1, all the code regarding this part in each ps1 was removed

juhlee-microsoft replied:
First, 'RunTest.psm1' should be license lines. 

There is a function Run-Test() in Libraries/RunTest.psm1, and it takes $DeployVM as parameter.
This function Run-Test is set inside AzureTestSuite.ps1. But it sets '$true' unconditionally to $deployVM if it is Azure. It does not make sense. It should check if VM is available. 

Also I mentioned this before. Change PS file name from RunTest.psm1 to the other. It will make people confuse with top level RunTests.ps1.

adriansuhov commented on Testscripts/Linux/verify-ifdown-ifup-eth0-rmmod-modprobe-hv-netvsc.sh:
```diff
@@ -14,67 +14,45 @@
+    TestIterations=1
 else
-		LogMsg "Setting Test Iterations to $TestIterations from constants.sh"
+    LogMsg "Setting Test Iterations to $TestIterations from constants.sh"
```
UtilsInit function from utils.sh does this now.

adriansuhov replied:
constants.sh is sourced by UtilsInit function in utils.sh

juhlee-microsoft replied:
OK

adriansuhov commented on Testscripts/Linux/nested_kvm_basic_test.sh:
```diff
@@ -63,30 +74,15 @@ fi
     if [ $exit_status -ne 0 ]; then
-        LogErr "Install KVM fail"
-        ResultLog  "Aborted"
+        LogMsg "Install KVM fail"
```
LogMsg is now a function used from utils.sh. Also "SetTestStateAborted/Failed/Completed" is used from there

lizzha replied:
LogMsg in utils.sh is used instead

juhlee-microsoft replied:
OK

adriansuhov commented on Testscripts/Windows/RHUI-BVT-VERIFY-RPM-SOURCE.ps1:
```diff
@@ -1,70 +1,49 @@
-$CurrentTestResult = CreateTestResultObject
-$resultArr = @()
-
-$isDeployed = DeployVMS -setupType $currentTestData.setupType -Distro $Distro -xmlConfig $xmlConfig
```
ditto

adriansuhov commented on Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1:
```diff
@@ -14,74 +20,61 @@ function Main {
-            if ($paramName -eq "rebootNumber") {
-                $rebootNumber = $paramValue
-            }
+    foreach ($param in $currentTestData.TestParameters.param) {
```
ditto

juhlee-microsoft replied:
Sometimes testing VM stuck at the booting step, and automation wait for infinitely. (We are testing against new stuff which can be unstable)

adriansuhov commented on Testscripts/Windows/VERIFY-TEST-SCRIPT-IN-LINUX-GUEST.ps1:
```diff
@@ -1,77 +0,0 @@
-﻿# Copyright (c) Microsoft Corporation. All rights reserved.
```
I can confirm. All the tests from xml files were run on Azure with the updated xml/testscripts folders as in this PR

juhlee-microsoft replied:
OK

ader1990 commented on AzureTestSuite.ps1:
```diff
@@ -187,6 +188,20 @@ Function RunTestsOnCycle ($cycleName , $xmlConfig, $Distro, $TestIterations )
+			$deployVM = $true
+			$removeVM = $true
+		} elseif ($testPlatform -eq "Hyperv") {
+			if ($testIndex -eq 1) {
```
Explain why you are doing this

ader1990 commented on AzureTestSuite.ps1:
```diff
@@ -241,11 +255,11 @@ Function RunTestsOnCycle ($cycleName , $xmlConfig, $Distro, $TestIterations )
 						LogMsg "Starting multiple tests : $($currentTestData.testName)"
 						$startTime = [Datetime]::Now.ToUniversalTime()
-						$CurrentTestResult = Invoke-Expression $command
+						$CurrentTestResult = Run-Test -CurrentTestData $currentTestData -XmlConfig $xmlConfig `
```
explain what this Run-Test method does from a black box to a white box perspective

ader1990 commented on Libraries/RunTest.psm1:
```diff
@@ -0,0 +1,338 @@
+function Get-SecretParams {
```
this method is named get-secretparams but inside is using testParams
why is that?

ader1990 commented on Libraries/RunTest.psm1:
```diff
@@ -0,0 +1,338 @@
+        [String] $Password,
+        [String] $TestType
+    )
+    # Note (mbivolan): This is a temporary solution untill a standard is decided
```
until*

ader1990 commented on Libraries/RunTest.psm1:
```diff
@@ -0,0 +1,338 @@
+    )
+    
+    foreach ($VM in $VMData) {
+        RemoteCopy -upload -uploadTo $VM.PublicIP -port $VM.SSHPort `
```
what happens if something  fails or timesout?

ader1990 commented on Libraries/RunTest.psm1:
```diff
@@ -0,0 +1,338 @@
+    
+    foreach ($VM in $VMData) {
+        $publicIP = ""
+        while (-not $publicIP) {
```
use an execute retry pattern with limited number of tries

ader1990 commented on Libraries/RunTest.psm1:
```diff
@@ -0,0 +1,338 @@
+    )
+    
+    $newVMData = @()
+    $publicIPRegex = "\b(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}\b"
```
use a .net class for this check

you can do a cast: [IPAddress] $ip

ader1990 commented on Libraries/RunTest.psm1:
```diff
@@ -0,0 +1,338 @@
+        while (-not $publicIP) {
+            LogMsg "    $VMname : Waiting for IP address..."
+            $vmNic = Get-VM -Name $VM.RoleName | Get-VMNetworkAdapter
+            $vmIP = $vmNic.IPAddresses | Where-Object {$_ -imatch $publicIPRegex}
```
what happens if the vm has more than one nic and each has more than one ip?

ader1990 commented on Libraries/RunTest.psm1:
```diff
@@ -0,0 +1,338 @@
+    $testPlatform = $XmlConfig.CurrentTestPlatform
+    $testResult = $false
+    
+    if ($DeployVM) {
```
get rid of the ifs here and use an inverse dependency using functional programming or classes

ader1990 commented on Testscripts/Linux/BVT-VERIFY-VERBOSE-ENABLED-LOGS.py:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#! /usr/bin/env python
```
for all the Python scripts run pep8 and flake8

juhlee-microsoft commented on Testscripts/Linux/core_stress_reload_modules.sh:
```diff
@@ -0,0 +1,171 @@
+#
+################################################################
+
+dos2unix utils.sh
```
Sure, Then here is my suggestion. Before running this step, check if dos2unix file exists or not. it can be downloaded if it does not exist. I try to remove binary from git repo and use blob contents in run-time.

iamshital replied:
@mbivolan also, using dos2unix requires the package should be installed beforehand.

ader1990 commented on AzureTestSuite.ps1:
```diff
@@ -148,6 +148,12 @@ Function RunTestsOnCycle ($cycleName , $xmlConfig, $Distro, $TestIterations )
 	$testsuite = StartLogTestSuite "CloudTesting"
+	
+	if ($testPlatform -eq "Azure") {
+		$DeployVMPerEachTest = $true
```
ExecuteSetupForEachTest

ader1990 commented on AzureTestSuite.ps1:
```diff
@@ -186,19 +192,10 @@ Function RunTestsOnCycle ($cycleName , $xmlConfig, $Distro, $TestIterations )
-				$removeVM = $true
-			}
+		if ($testIndex -eq 1) {
+			$TestState = @{"IsFirstTest" = $true}
```
remove all the skipped tests before looping and executing them

ader1990 replied:
replace isfirsttest with ExecuteSetup

ader1990 commented on AzureTestSuite.ps1:
```diff
@@ -186,19 +192,10 @@ Function RunTestsOnCycle ($cycleName , $xmlConfig, $Distro, $TestIterations )
+		if ($testIndex -eq 1) {
+			$TestState = @{"IsFirstTest" = $true}
+		} elseif ($testIndex -eq $testCount) {
+			$TestState = @{"IsLastTest" = $True}
```
ExecuteTeardown

iamshital commented on Libraries/Test-Framework.psm1:
```diff
@@ -0,0 +1,499 @@
+            Set-Variable -Name isDeployed -Value $isDeployed -Scope Global
+        }
+    } else {
+        if ($testPlatform.ToUpper() -eq "HYPERV") {
```
The current implementation **strictly** uses checkpoint method for HyperV mode. However, we're missing a key functionality here - First Boot. In Azure, we have an option called as - EconomyMode. Which decides, whether to run test on Same VMs again and again, or to create new VMs for each test. Nowadays, EconomyMode is not being used because, many times, we've found the boot related issues in First Boot. In conjunction with "Checkpoint" method, I strongly recommend to re-enable "New VMs" approach for HyperV tests as well. And let the user decide, how it wants to run the tests.

adriansuhov replied:
We're leaving this now as it is because it's not affecting the functionality. Once this is merged, we can talk about adding this feature

mbivolan commented:
Review Changes:
- @juhlee-microsoft:
	- Keep bvt names consistent
	- Change Licence
	- Added exception line logging in powershell test scripts
	- Removed dos2unix command and other refactors for shell test scripts
	- Small refactor for python scripts
- @lizzha:
	- Refactored Check-IP method
	- Refactored Enable-RootUser
- @ader1990:
	- Added timeout in Check-IP method
	- Changed variable names in AzureTestSuite
	- Moved Test Skip before test looping

adriansuhov commented on Testscripts/Windows/VERIFY-NVIDIA-CUDA-DRIVER-INSTALLATION.ps1:
```diff
@@ -1,209 +1,166 @@
-		#region Install N-Vidia Drivers and reboot.
-		$myString = @"
+param (
+    [String] $VmName,
```
WIll remove the parameters for now

adriansuhov commented on Testscripts/Windows/BVT-VERIFY-KVP-DAEMON-STATUS.ps1:
```diff
@@ -1,74 +1,50 @@
-		
-	}
+param (
+    [String] $VmName,
```
ditto

adriansuhov commented on Testscripts/Windows/BVT-VERIFY-DEPLOYMENT-PROVISION.ps1:
```diff
@@ -1,74 +1,50 @@
+    [String] $VmName,
+    [String] $VmLocation,
+    [string] $TestParams
+)
```
ditto

adriansuhov commented on Testscripts/Windows/BVT-VERIFY-PRIVATEKEY-AUTHENTICATION.ps1:
```diff
@@ -1,85 +1,64 @@
+    [String] $VmName,
+    [String] $VmLocation,
+    [string] $TestParams
+)
```
ditto

iamshital commented on Testscripts/Windows/FILE-SYSTEM-VERIFICATION-TESTS.ps1:
```diff
@@ -1,85 +1,71 @@
+        foreach ($param in $currentTestData.TestParameters.param) {
+            if ($param -imatch "FSTYP=") {
                 $TestFileSystem = ($param.Replace("FSTYP=",""))
                 Add-Content -Value "[$TestFileSystem]" -Path $constantsFile
```
XFS tests are broken.
This is line is missing from the constants.sh file which is being prepared as part of test framework

mbivolan replied:
This has nothing to do with constants.sh file. The test scripts generates a .config file (xfstests-config.config) which is still uploaded and used in the test. Plus if we would append "[$TestFileSystem]" in a shell script, sourcing it would fail, as "[$TestFileSystem]" is not a valid command.

iamshital replied:
Confirmed

mbivolan commented on Libraries/Test-Framework.psm1:
```diff
@@ -0,0 +1,499 @@
+    }
+}
+
+function Check-IP {
```
GetAllHyperVDeployementData does a lot more that just IP checking, all we need here is just to check for IP after VM restart / checkpoint restore.
